azure/services/natgateways/spec.go

```diff
@@ -19,81 +19,87 @@ package natgateways
+// Parameters returns the parameters for the NAT gateway. It implements azure.ASOResourceSpecGetter.
+func (s *NatGatewaySpec) Parameters(ctx context.Context, existing *asonetworkv1.NatGateway) (params *asonetworkv1.NatGateway, err error) {
+	// no need to set ObjectMeta as it is already set in ResourceRef
+	natGatewayToCreate := &asonetworkv1.NatGateway{}
 	if existing != nil {
-		existingNatGateway, ok := existing.(armnetwork.NatGateway)
-		if !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.NatGateway", existing)
-		}
-
-		if hasPublicIP(existingNatGateway, s.NatGatewayIP.Name) {
-			// Skip update for NAT gateway as it exists with expected values
-			return nil, nil
+		if hasPublicIP(existing, s.NatGatewayIP.Name) {
+			// return existing NAT gateway as it exists with expected values
+			return existing, nil
 		}
+		// no need for deepcopy as "existing" is already a deep copy
+		natGatewayToCreate = existing
 	}
 
-	natGatewayToCreate := armnetwork.NatGateway{
-		Name:     ptr.To(s.Name),
+	natGatewayToCreate.Spec = asonetworkv1.NatGateway_Spec{
+		Owner: &genruntime.KnownResourceReference{
+			Name: s.ResourceGroup,
+		},
 		Location: ptr.To(s.Location),
-		SKU:      &armnetwork.NatGatewaySKU{Name: ptr.To(armnetwork.NatGatewaySKUNameStandard)},
-		Properties: &armnetwork.NatGatewayPropertiesFormat{
-			PublicIPAddresses: []*armnetwork.SubResource{
-				{
-					ID: ptr.To(azure.PublicIPID(s.SubscriptionID, s.ResourceGroupName(), s.NatGatewayIP.Name)),
+		Sku: &asonetworkv1.NatGatewaySku{
+			Name: ptr.To(asonetworkv1.NatGatewaySku_Name_Standard),
+		},
+		PublicIpAddresses: []asonetworkv1.ApplicationGatewaySubResource{
+			{
+				Reference: &genruntime.ResourceReference{
+					ARMID: azure.PublicIPID(s.SubscriptionID, s.ResourceGroup, s.NatGatewayIP.Name),
 				},
 			},
 		},
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+		Tags: infrav1.Build(infrav1.BuildParams{
 			ClusterName: s.ClusterName,
 			Lifecycle:   infrav1.ResourceLifecycleOwned,
 			Name:        ptr.To(s.Name),
 			Additional:  s.AdditionalTags,
-		})),
+		}),
 	}
 
 	return natGatewayToCreate, nil
 }
 
-func hasPublicIP(natGateway armnetwork.NatGateway, publicIPName string) bool {
-	for _, publicIP := range natGateway.Properties.PublicIPAddresses {
-		if publicIP != nil && publicIP.ID != nil {
-			resource, err := azureutil.ParseResourceID(*publicIP.ID)
+// WasManaged implements azure.ASOResourceSpecGetter.
+func (s *NatGatewaySpec) WasManaged(resource *asonetworkv1.NatGateway) bool {
+	return s.IsVnetManaged
+}
+
+// hasPublicIP returns true if the public IP Name is already attached to the NAT gateway.
+func hasPublicIP(natGateway *asonetworkv1.NatGateway, publicIPName string) bool {
+	for _, publicIP := range natGateway.Status.PublicIpAddresses {
+		if publicIP.Id != nil {
+			resource, err := azureutil.ParseResourceID(*publicIP.Id)
```

Comment:
I'm wondering if it might be preferable here to use `Spec` instead of `Status` (I know I said otherwise earlier). If the IP gets removed from the spec for some reason and ASO starts removing it, `Status` can be out of date and still have it, and this function would return `true`. Then we'd have to wait for ASO to converge, pick up that change, then notice that it's gone from the status to add it back to the spec. I think if we check the spec directly, we would basically skip waiting for ASO to converge on the removed IP before adding it back.

I'm not sure if that exact scenario is possible here, but I think for the sake of trying to establish some rules of thumb while we convert the other services it's worth thinking about.

```suggestion
	for _, publicIP := range natGateway.Spec.PublicIpAddresses {
		if publicIP.Reference != nil {
			resource, err := azureutil.ParseResourceID(publicIP.Reference.ARMID)
```

Reply:
I think sticking to ASOResource’s `status` is safer than its `spec`. 

I agree that there will be a delay in propagating the change from `spec` to `status`; but it is _safer_, since CAPZ is relying on ASO to perform the background jobs of whatever it needs to do to reflect the changes in the the`spec` of the resource -> reconcile them -> and then show it in the resource’s `status`.  While grabbing _the detail(s)_ from the `spec` solves the delay issue, it can push the resource in weird scenarios where CAPZ tries to perform duplicate `PATCH` onto the K8s resource.

Reply:
I'm not sure I understand why using spec could be less safe here or produce more duplicate work than using status. If we check the status, then there's a chance that the spec is already up to date in which case we might try to issue an update even though the spec is already up to date.

Reply:
I think the conclusion we came to offline with @jackfrancis and @willie-yao here is that this check (and similar ones) shouldn't even be necessary at all in the context of ASO. Each `Parameters()` implementation should more-or-less blindly construct the `spec` of an ASO resource and leverage the higher-level `CreateOrUpdateResource()` defined in the `aso` package to skip sending updates to ASO if they aren't required.

Before with the SDK, that "skip update if not modified" logic had to be defined in each service because any non-nil value returned by `Parameters()` would always result in an Azure API call.

Reply:
Thank you for summarizing @nojnhuh !

Reply:
This should be now updated via c4dbab91d193e43e4d261c67f8414e3147f8cfa7